src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -338,32 +338,82 @@ internal static string ToDebugString(this BindingRestrictions restrictions)
 
     internal static class DynamicMetaObjectBinderExtensions
     {
-        internal static DynamicMetaObject DeferForPSObject(this DynamicMetaObjectBinder binder,
-                                                           params DynamicMetaObject[] args)
+        internal static DynamicMetaObject DeferForPSObject(this DynamicMetaObjectBinder binder, DynamicMetaObject arg0, bool moreRestrictionForArg0 = false)
+        {
+            Diagnostics.Assert(arg0.Value is PSObject, "arg1 must be a psobject");
+
+            BindingRestrictions restrictions = BindingRestrictions.Empty;
+            Expression expr = ProcessOnePSObject(arg0, ref restrictions, checkBaseTypeForRestriction: moreRestrictionForArg0);
+            return new DynamicMetaObject(DynamicExpression.Dynamic(binder, binder.ReturnType, expr), restrictions);
+        }
+
+        internal static DynamicMetaObject DeferForPSObject(this DynamicMetaObjectBinder binder, DynamicMetaObject arg0, DynamicMetaObject arg1, bool moreRestrictionForArg0 = false)
+        {
+            Diagnostics.Assert(arg0.Value is PSObject || arg1.Value is PSObject, "At least one arg must be a psobject");
+
+            BindingRestrictions restrictions = BindingRestrictions.Empty;
+            Expression expr1 = ProcessOnePSObject(arg0, ref restrictions, checkBaseTypeForRestriction: moreRestrictionForArg0);
+            Expression expr2 = ProcessOnePSObject(arg1, ref restrictions, checkBaseTypeForRestriction: false);
+            return new DynamicMetaObject(DynamicExpression.Dynamic(binder, binder.ReturnType, expr1, expr2), restrictions);
+        }
+
+        internal static DynamicMetaObject DeferForPSObject(this DynamicMetaObjectBinder binder, DynamicMetaObject[] args, bool moreRestrictionForArg0 = false)
         {
             Diagnostics.Assert(args.Any(mo => mo.Value is PSObject), "At least one arg must be a psobject");
 
             Expression[] exprs = new Expression[args.Length];
             BindingRestrictions restrictions = BindingRestrictions.Empty;
             for (int i = 0; i < args.Length; i++)
             {
-                var baseValue = PSObject.Base(args[i].Value);
-                if (baseValue != args[i].Value)
+                bool checkBaseTypeForRestriction = false;
+                if (i == 0) { checkBaseTypeForRestriction = moreRestrictionForArg0; }
```

Comment:
We could exclude this by call overload `DeferForPSObject` if `args.Length == 1` before the cycle.